### ISSUE TITLE
Running recipes from parent directories when there's no recipe with the same name in the current directory

### ISSUE THREAD

**Issue by soenkehahn** [casey/just#1036]
When my current directory does not contain a `justfile` but a parent directory does, `just` looks for recipes in the `justfile` in that parent dir. This allows me to e.g. place a `justfile` in the root of a monorepo and have its recipes available everywhere.

However, this does not work when my current directory contains a `justfile`, even if that `justfile` does not contain the recipe I want to run. Here's an example:

``` bash
$ cat justfile
global-recipe:
  echo global-recipe
$ cd subdir/
$ cat justfile
local-recipe:
  echo local-recipe
$ just local-recipe
echo local-recipe
local-recipe
$ just global-recipe
error: Justfile does not contain recipe `global-recipe`.
Did you mean `local-recipe`?
$
```
I would've expected this to work, as long as my local `justfile` doesn't shadow the recipe that's higher up. Is there a good reason why this shouldn't work?

**Comment by casey** [comment]
My initial reaction is that a module system is the way to support this. See #383 and #929.

I can imagine an implementation that would use the first recipe to inform how just searches for the justfile, but I'm not sure if that would favorably compare to a module system.

A module system is a huge new feature, but maybe it could be scoped down to the point where the initial implementation isn't too bad, and it supports this use-case. There's also the `--unstable` flag, which can be used to add incomplete features that can change in backwards incompatible ways. We could also have an unstable setting, i.e. `set unstable` that could do this without needing to pass a flag.

**Comment by asdf8dfafjk** [comment]
Since you consider module system to be huge, and since a rudimentary form of inheritance works right now, I wonder if #761  (pardon the possibly rude repetition) might fit? I'd be happy to tackle the actual work btw.

**Comment by casey** [comment]
I think the first version of the module system could be really simple. Start in the current directory, and as you go up to the justfile, every time you see a file called "mod.just", add that to a list. Then, when you get to the justfile, parse the justfile, and all the modules in the list.

Then, start looking in the justfiles one by one, starting with the deepest going upwards. When you find a justfile with the recipe, use that justfile.

**Comment by asdf8dfafjk** [comment]
Yeah, that's actually pretty easy to mentally model.

**Comment by casey** [comment]
There are some tricky things to consider here. One example is, if the recipe isn't in the current justfile, and you search up for another justfile, and you find another justfile but there's a parse error, do you display that error? I started implementing this, but hit some of these tricky things and stopped working on it. I think now I should have just finished it, and put it behind the `--unstable` flag, and addressed the tricky bits later.

**Comment by soenkehahn** [comment]
> [...] do you display that error?

I do agree that this is a bit tricky, but I would assume that yes, you just display that error.

I guess another question is whether `just --list` shows recipes from justfiles in parent directories or not. I would've assumed that it doesn't, but it's weird that you can execute a recipe that isn't in that list.

**Comment by eelkevdbos** [comment]
> There are some tricky things to consider here. One example is, if the recipe isn't in the current justfile, and you search up for another justfile, and you find another justfile but there's a parse error, do you display that error? I started implementing this, but hit some of these tricky things and stopped working on it. I think now I should have just finished it, and put it behind the `--unstable` flag, and addressed the tricky bits later.

First of all, this would be a killer feature @casey @soenkehahn. Some of the solutions that come to mind:

- Have some stack tracing available on a parse error that highlights which file is currently being parsed, this could be a nice addition as well to the error 'No justfiles found' to see which paths are walked
- Add an option `--verbose/--debug` that allows for more verbosity, printing the actual parsing steps as debug logs to stdout
- Add an option `--no-modules` to disable module loading behavior, and provide the use of this option in the stack trace mentioned above to help people debug their problem.
- Add an option to specifically include modules `--module "ops" --module "docker"`, this could also be useful in context of the remote modules talked about in the other issues `--module "https://github.com/org/repo/tree/main/docker.just"`

Other things that come to mind:
- Being very explicit in the most specific `.justfile` that expansion is allowed / disallowed. This could even be used to limit expansion to arbitrary levels: `set allowed_modules = '../'` or `set allowed_modules = ['~/', '../', 'https://github.com/org/repo/tree/main/docker.just']`

**Comment by casey** [comment]
I implemented a simple version of this feature in #1149 and released a new version of just, 1.1.2, which contains it. Try it out and let me know what you think!

The feature is currently behind the `--unstable` flag, so I created #1148 to track feedback prior to stabilization.

**Comment by hrumhurum** [comment]
The documentation for the implemented fallback feature (requires `just` version 1.11.0+):

"Fallback through a `justfile` in which a recipe is not found will only occur if the `justfile` contains `set fallback := true` directive."

( just for the reference of people who stumble upon this thread)
